text/0523-model-argument-for-route-templates.md

```diff
@@ -0,0 +1,114 @@
+a JavaScript class was needed in the past (e.g. to customize the "wrapper"
+element) can now be expressed with templates alone. This is a direction we will
+continue post-Octane.
+
+We would like to update the learning materials (guides) to focus on teaching
+the template-centric component model. For example, components can be introduced
+as a way to break up large templates into smaller, named pieces, similar to
+refactoring a big function into smaller ones. Then, we can layer on making them
+reusable through passing arguments. Next, we can introduce the idea of passing
+a block and yielding. Finally, we can introduce the component class when we are
+ready to discuss adding behavior to the component.
+
+As you can see, we can accomplish quite a lot with templates-only components in
+Octane, and focusing on the teaching template first-and-foremost would provide
+a gentle learning curve for developers and designers who are comfortable with
+HTML but perhaps new to Ember (or even JavaScript). With this flow, the concept
+of a component class, and consequently, the idea of a "`this` context" in a
+template comes up quite late.
+
+This presents a problem, as we may want or need to teach route templates before
+that was introduced. As the model can only be accessed through `this.model` in
+a route template, we would be forced to introduce that concept (and the contept
+of a controller) at an earlier time than would be ideal.
+
+Providing `@model` to route templates would solve this problem quite elegantly.
+We will now be able to introduce the concept of arguments (`@name`) once and it
+can be applied consistently between component and route templates.
+
+This also aligns with the general mental model that arguments are things that
+are passed into the template from the outside (which is true in the case of the
+route model).
+
+This can also be thought of as a small incremental step in the bigger picture
+of reforming route templates and removing controllers from Ember. Specifically,
+it moves us a bit closer to the mental model that controllers/route templates
+are just a "special" kind of component. We expect to continually unify them and
+remove the remaining differences, and this is a step towards that direction.
+
+## Detailed design
+
+Internally, route templates are _already_ modelled as components at the Glimmer
+VM layer. To implement this, we would "synthesize" a named argument `@model`
+containing the resolved route model, i.e. the same value as `this.model` on the
+controller instance.
+
+Just like the "reflected" named arguments in classic components, mutating
+`this.model` on the controller instance would _not_ change the value of
+`@model`. In practice, this is unlikely to be relied upon as mutating the
+model property on the controller is considered a bad practice and is not really
+supported (e.g. `route.modelFor` would return the "wrong" value).
+
+## How we teach this
```

Comment:
I think there is one very important unanswered pedagogical question here: what *is* this template? Is it or is it not a component? (It seems like it is *not* a component, but something *akin* to a component.)

If we simply teach it as "route templates are just a special kind of template-only component which receive `@model` as an argument", then the teaching model around controllers gets *weirder* in the timeframe between when we ship this and when we replace controllers with a true component-based model. Maybe that's fine, but we should make explicit here IMO.

Reply:
They are (in implementation terms, at least) "outer-html" components whose context/instance is the controller. You can think of `Ember.Controller` as a special kind of [custom components](https://emberjs.github.io/rfcs/0213-custom-components.html) with weird lifecycle semantics (they are singletons).

Whether we want to _teach_ them like that, or draw attention to that fact, is a separate question, but I do think the idea of "route templates are components" are internally consistent, because that's how they are actually implemented.

I think the singleton aspect is indeed weird, but really, that's weird no matter what (whether you think of them as components or not), and I think most of us agree we should figure out how to remove that "feature" sooner rather than later anyway.

Reply:
Very much agreed on the overall direction and implementation angle; my concern is *solely* on the pedagogical front here.

> You can think of a route template as being a special form of component. If it has no backing class, it is just like any other template-only component, and it receives a `@model` argument.
> 
> If it has a backing class, that class is a `Controller` instead of a `Component`. We'll cover those details in the `Controllers` section…

Something like that probably gets the job done, and I think it's probably *fine* in the interim. 🤔  (Not suggesting that for actual text, to be clear, just thinking out loud.)

Reply:
I think that's fine. In general, I think we don't really want/need to teach template-only components as a different thing than a "regular" component. It's more like "at its core, a component is just a named piece of template" -> "if you need behavior, you can add a JS file". I think that's applicable here too.

Reply:
Template only components don’t add the wrapping element though right? Just like the route templates will add no wrapping element.

Reply:
That's correct. However, neither do Glimmer components, which will be the default by the time this is implemented—both in that the feature flag for template-only-components behaving this way will be enabled by default _and_ that `@glimmer/component` has `outerHTML` semantics and *also* will be enabled by default for Octane apps.

Reply:
@webark @chriskrycho going forward, I think this is a better way to think about it:

1. At its very core, a component is just a named piece of template, and "what you type (in the template) is what you get (when invoking the component)". So, components, in general, are "outerHTML".

2. Sometimes (and maybe quite often), your component need to be _more_ than just _presentational_, so just a piece of template doesn't cut it in those cases. They need _behavior_, and in those cases, you can add a class for them.

3. Most of the time, the class should simply provides a blank canvas for your JavaScript (behavior/logic) code and mostly get out of your way. `@glimmer/component` is our attempt at that to provide an intentionally minimal API to get out of your way, and is a good default going forward.

4. However, Ember's component system is "pluggable". When `@glimmer/component` is not sufficient, or for whatever reason is not appropriate, you can inherit from a different super class that suits your needs, whether that is `@ember/component`, `@ember/controller`, or other [custom components](https://emberjs.github.io/rfcs/0213-custom-components.html) provided by an addon.

5. These super classes are free to provide different APIs than to your class that makes the task at hand easier. For example, `@ember/component` provides and API for using positional arguments, and `@glimmer/component` does not.

6. In additional to just changing the API though, these super classes are allowed to customize other low-level aspects of the rendering logic. For example, `@ember/components` adds the "wrapper element" around your template, and `@ember/controller` caches the instances and make them singleton. These examples are not particularly good uses of those low-level capabilities, and are probably undesirable, but nevertheless, these kind of things are what the super class gets to decide for you, which you are opting into when subclassing from them.

So, to recap:

* Components = Templates (+ a class, sometimes)
* Templates: what you type is what you get, no wrapper, no magic
* Component class: subclass from some super class
* Super class: has special powers

The main thing I wanted to point out is that, the "special behavior" (no wrapper element) is not in `@glimmer/component`. That should be the behavior, always, for all components. The thing that _is_ special is `@ember/component`. Template-only components are a special thing either. They are just the most basic form of a component, a component that happens to not have a class. Since they didn't subclass from anything, they don't get any special behavior either. So no wrapper element also.

The fact that, once upon a time, "template-only" component automatically get an instance of `@ember/component` was a mistake, and hopefully one that we will soon forget about going forward.